pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/SaxonXPathRuleQuery.java

```diff
@@ -206,15 +251,55 @@ private void initializeXPathExpression() {
                 }
             }
 
-            // TODO Come up with a way to make use of RuleChain. I had hacked up
-            // an approach which used Jaxen's stuff, but that only works for
-            // 1.0 compatibility mode. Rather do it right instead of kludging.
             xpathExpression = xpathEvaluator.createExpression(super.xpath);
+            analyzeXPathForRuleChain(xpathEvaluator);
         } catch (final XPathException e) {
             throw new RuntimeException(e);
         }
     }
+    
+    private void analyzeXPathForRuleChain(final XPathEvaluator xpathEvaluator) {
+        final Expression expr = xpathExpression.getInternalExpression();
+
+        boolean useRuleChain = true;
+
+        // First step: Split the union venn expressions into single expressions
+        List<Expression> subexpressions = new ArrayList<>();
+        SplitUnions unions = new SplitUnions();
+        unions.visit(expr);
+        if (unions.getExpressions().isEmpty()) {
+            subexpressions.add(expr);
+        } else {
+            subexpressions.addAll(unions.getExpressions());
+        }
 
+        // Second step: Analyze each expression separately
+        for (Expression subexpression : subexpressions) {
+            RuleChainAnalyzer rca = new RuleChainAnalyzer(xpathEvaluator.getConfiguration());
+            Expression modified = rca.visit(subexpression);
+
+            if (rca.getRootElement() != null) {
+                addExpressionForNode(rca.getRootElement(), modified);
+            } else {
+                // couldn't find a root element for the expression, that means, we can't use rule chain at all
+                // even though, it would be possible for part of the expression.
+                useRuleChain = false;
+                break;
+            }
+        }
+
+        if (useRuleChain) {
+            super.ruleChainVisits.addAll(nodeNameToXPaths.keySet());
+        } else {
+            nodeNameToXPaths.clear();
+            if (LOG.isLoggable(Level.FINE)) {
+                LOG.log(Level.FINE, "Unable to use RuleChain for XPath: " + xpath);
+            }
+        }
+
+        // always add fallback expression
+        addExpressionForNode(AST_ROOT, xpathExpression.getInternalExpression());
```

Comment:
I think this undermines the benefits of using the rulechain? We'd effectively traverse once for each rulechain visit, and once for the whole tree, so we'd only be doing *more* traversals, while the whole tree visit suffices

Reply:
That should be the case - but I double check.

As far as I know: `getRuleChainVisits()` only returns the nodes, that should be visited and the rule chain calls the rule then only for these nodes. If it happens, that the rule called for whatever reason on a different node, that we would use the whole expression.
If we couldn't split up the expression, then the getRuleChainVisits() will return a empty list and for java the rule is called with ASTCompilationUnit - since we don't have a specific expression for this node, we would execute the whole expression (aka. fallback - AST_ROOT).

Actually, the current master impl returns the node name "_AST_ROOT" also in the "getRuleChainVisits()" list - luckily, we have no node named "_AST_ROOT", so no extra nodes are visited...
See https://github.com/pmd/pmd/pull/2354/commits/5fb6df7daa5d0eb36e963d8f4eef1853444b759f#diff-dd6490e8d37457789b60964710470e15R46 - I didn't fix that bevior for jaxen, because it should not matter.

---
When executing via rulechain, we get called here
https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleChainVisitor.java#L87-L99
For each node separately.

This eventually ends up in the SaxonQuery, where we get the expressions for the node:
https://github.com/pmd/pmd/pull/2354/files#diff-ae9e480e9dfbc219210827dca4d4a6d8R132-R137

If there is no specific expression, we would return here the expression for AST_ROOT. But we only get called for the nodes, we explicitly returned via getRuleChainVisits()...

So, I don't see a problem here....

Reply:
Ok, thanks for explaining. Then, I think we can merge this and see how it goes. I'll update the release notes when merging